### PR TITLE
Inotify backend tests

### PR DESCRIPTION
**Pull request by aalekseyev**
Refactor dune file watcher backend and add some tests.

In particular:

- move the file watcher to its own library: dune_file_watcher
- add a few new functions to its .mli, as driven by the need to make tests reproducible. In particular `wait_watches_established_blocking` will be needed in production so that we subscribe to notifications before we start reading files.
- add a few tests, demonstrating some bugs.

After doing this, I'm becoming more skeptical that inotifywait is a good backend: it does not protect us from the complexities of inotify and contributes its own bugs, so we should probably use inotify API directly.

### CODE REVIEW

File: src/dune_file_watcher/dune_file_watcher.ml [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+  ; wait_for_watches_established : unit -> unit
+  }
+
+let pid t = t.pid
+
+let buffer_capacity = 65536
+
+(* Fixed-size buffer for reading line-by-line from file descriptors. Bug:
+   deadlocks if there's a line longer than the capacity of the buffer. TODO: use
+   Out_channel? *)
```
Comment (jeremiedimino) [inline src/dune_file_watcher/dune_file_watcher.ml:14]:
```suggestion
   In_channel? *)
```

Comment (aalekseyev) [reply src/dune_file_watcher/dune_file_watcher.ml]: Thanks. Fixed.

Comment (jeremiedimino) [reply src/dune_file_watcher/dune_file_watcher.ml]: Using `in_channel` seems fine. IIRC, we tried that initially and had some issues, maybe related to blocking. But the code was a bit different, in particular we didn't have the debouncing thread. So maybe we can make it work. In any case, 64k should be fine on most systems given the max path length.

**Review by jeremiedimino — CHANGES_REQUESTED** [review]
Looks good, but we should probably sandbox the tests.

File: test/expect-tests/dune_file_watcher/dune_file_watcher_tests.ml [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+          Some list)
+  in
+  let print_events n = print_events ~try_to_get_events ~expected:n in
+  Dune_file_watcher.wait_watches_established_blocking watcher;
+  Stdio.Out_channel.write_all "x" ~data:"x";
+  print_events 1;
+  [%expect {|
+x
+|}];
+  (* CR-someday aalekseyev: renaming is not detected *)
```
Comment (jeremiedimino) [inline test/expect-tests/dune_file_watcher/dune_file_watcher_tests.ml:110]: :flushed: never realised we were ignoring all these events. Seems we are not passing enough `--event` arguments to  inotifywait.

File: test/expect-tests/dune_file_watcher/dune [hunk 3]
```diff
@@ -0,0 +1,16 @@
+(library
+ (name dune_file_watcher_tests)
+ (inline_tests (deps))
```
[the remark this answers is not in the document]
Comment (aalekseyev) [reply test/expect-tests/dune_file_watcher/dune:3]: Good point! Fixing.

**Review by jeremiedimino — APPROVED** [review]

File: test/expect-tests/dune_file_watcher/dune_file_watcher_tests.ml (hunk 2 above)
Comment (aalekseyev) [reply test/expect-tests/dune_file_watcher/dune_file_watcher_tests.ml:110]: Yeah, adding a bunch of `--event` arguments could definitely help.

File: test/expect-tests/dune_file_watcher/dune_file_watcher_tests.ml [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+  Unix.rename "x" "y";
+  print_events 0;
+  [%expect {|
+|}];
+  Dune_file_watcher.For_tests.suspend watcher;
+  let (_ : _) = Fpath.mkdir_p "d/w" in
```
Comment (snowleopard) [inline test/expect-tests/dune_file_watcher/dune_file_watcher_tests.ml:116]: `(_ : _)` seems simplifiable to just `_` :)

Comment (snowleopard) [reply test/expect-tests/dune_file_watcher/dune_file_watcher_tests.ml:116]: (Not only I won't tell you the name, I explicitly refuse to tell you the type either!)

**Review by snowleopard — APPROVED** [review]
Looks great!

File: test/expect-tests/dune_file_watcher/dune_file_watcher_tests.ml (hunk 4 above)
Comment (aalekseyev) [reply test/expect-tests/dune_file_watcher/dune_file_watcher_tests.ml:116]: You laugh, but I think that's how you have to write it to appease the js style checks, so it's in my muscle memory now. :-)

Comment (snowleopard) [reply test/expect-tests/dune_file_watcher/dune_file_watcher_tests.ml:116]: Oh, I see, good to know :)

**Pull request merged**